packages/daemon/src/remote-control.js

```diff
@@ -0,0 +1,227 @@
+// @ts-check
+
+/**
+ * @param {string} localNodeId
+ */
+export const makeRemoteControlProvider = localNodeId => {
+  /** @type {Map<string, import('./types.js').RemoteControl>} */
+  const remoteControls = new Map();
+
+  /** @param {string} remoteNodeId */
+  const makeRemoteControl = remoteNodeId => {
+    /** @type {import('./types.js').RemoteControlState} */
+    let state;
+
+    // In this state, we have received a remoteGateway from an ingress
+    // connection (and provided our local gateway to them.)
+    // We do not have a pending outbound connection attempt.
+    /**
+     * @type {(
+     *   remoteGateway: Promise<import('./types.js').EndoGateway>,
+     *   cancel: (error: Error) => void,
+     *   cancelled: Promise<never>,
+      ) => import('./types.js').RemoteControlState} */
+    const accepted = (remoteGateway, cancelCurrent, currentCancelled) => {
+      return {
+        accept(_proposedRemoteGateway, proposedCancel, _proposedCancelled) {
+          // And we receive an inbound connection.
+          // There are two possibilities:
+          // The sender raced multiple outbound connections and this connection
+          // lost the race.
+          // Or, the sender restarted but left a prior connection half-open.
+          // We consider the race to be the common consideration and that
+          // replacing peers after establishing a connection would be far to
+          // disruptive.
+          // TODO: For the case where we leave a peer wedged half-open, we
+          // will need health checks.
+          proposedCancel(new Error('Already accepted a connection.'));
+          return accepted(remoteGateway, cancelCurrent, currentCancelled);
+        },
+        connect(_getProposedRemoteGateway, proposedCancel, proposedCancelled) {
+          // Use the gateway we already have.
+          // Bind the fates of the current peer incarnation and the inbound
+          // connection.
+          currentCancelled.catch(proposedCancel);
+          proposedCancelled.catch(cancelCurrent);
+          return {
+            state: accepted(remoteGateway, proposedCancel, proposedCancelled),
+            remoteGateway,
+          };
+        },
+      };
+    };
+
+    // We have an active outbound connection.
+    /**
+     * @type {(
+     *   remoteGateway: Promise<import('./types.js').EndoGateway>,
+     *   cancel: (error: Error) => void,
+     *   cancelled: Promise<never>,
+      ) => import('./types.js').RemoteControlState} */
```

Comment:
Same formatting issue as above:
```suggestion
     *   cancelled: Promise<never>,
     * ) => import('./types.js').RemoteControlState}
     */
```